Adds completely new Reference section, including Components. (#127)
shinzonetwork/docs: content/reference/architecture-overview/index.mdx

---
title: "Architecture overview"
---

Shinzo has two layers: an on-chain coordination layer and an off-chain data layer. The on-chain layer handles registration, payments, and access control. The off-chain layer handles the actual blockchain data, from fetching it through signing and transforming it to serving it to users.

Neither layer works without the other. The on-chain layer controls who can participate. The off-chain layer does the grunt-work.

> ## System diagram

Reviewer: Great diagram

```mermaid
flowchart LR
  Src["Source chains<br/>(Ethereum, Cosmos, …)"]

  subgraph OffChain["Off-chain data layer"]
    direction TB
    Idx["Indexer<br/>(Geth + indexer client)"]
    Host["Host<br/>(host client + DefraDB)"]
    Idx -- "signed blocks" --> Host
  end

  subgraph OnChain["On-chain coordination layer"]
    direction TB
    SH["ShinzoHub<br/>(registries, payments)"]
    Source["SourceHub<br/>(ACP / access control)"]
    SH <-- "ICA / IBC" --> Source
  end

  subgraph Coord["Coordination services"]
    direction TB

>     Sched["Scheduler"]

Reviewer: This is connected to shinzohub in order to get the registrations

:::

## Off-chain data flow

### DefraDB

DefraDB is the peer-to-peer document database embedded in every indexer and host. Three separate DefraDB instances exist in the system (one per indexer, one per host for primitives, one per host for view data).

DefraDB uses MerkleCRDTs, a combination of Merkle DAGs and CRDTs. Document updates are stored as a Merkle DAG where each node is a CRDT operation. CRDTs give you conflict resolution without coordination. The Merkle DAG gives you verifiable history. And because nodes can diff DAGs, sync only transfers the missing pieces.

Documents are content-addressed using CIDs (Content Identifiers). The CID is a hash of the document content, so verification is just re-hashing and comparing. DefraDB schemas are defined in GraphQL SDL, and queries use standard GraphQL syntax.

### P2P topology

Data replication uses libp2p, managed internally by DefraDB. The flow for a new document:

1. Indexer writes document to its local DefraDB.

> 1. DefraDB computes a content digest.

Reviewer: ```suggestion
2. DefraDB computes a content digest.
```

Author: The markdown parser automatically converts `1.` to whatever number it needs to be. If we manually put in the numbers (like this suggestion) then each time we add a new item into the list, we have to update all the numbers manually.

> 1. DefraDB gossips the digest to connected peers.

Reviewer: ```suggestion
3. DefraDB gossips the digest to connected peers.
```

Author: https://github.com/shinzonetwork/docs/pull/127/changes#r3194455042

> 1. Peers that want the document request the full content.

Reviewer: ```suggestion
4. Peers that want the document request the full content.
```

Author: https://github.com/shinzonetwork/docs/pull/127/changes#r3194455042

> 1. DefraDB sends the full document.

Reviewer: ```suggestion
5. DefraDB sends the full document.
```

Author: https://github.com/shinzonetwork/docs/pull/127/changes#r3194455042

Indexers are write-only. They publish documents and reject all incoming replication (enforced by a replication filter in the indexer client). Hosts accept incoming documents from indexers and replicate attestation records between each other.

Peers discover each other through bootstrap peers configured in DefraDB and through `EntityRegistered` events from ShinzoHub (when new indexers or hosts join the network).

### Signing and attestation

> Each indexer signs every block batch it produces. After writing all documents for a block (Block, Transaction, Log, AccessListEntry), the indexer computes a Merkle root over all document CIDs, signs it with its identity key, and writes a `BlockSignature` document. Individual documents also carry per-document signatures in their `_version` array.

Reviewer: The indexer client creates a function specific document called a `blockSignature`. This document is a collection of document metadata, block specific fields, and cryptographic fields. The indexer computes a merkle root for the entire list of CIDs, signs it and writes it locally.  The host then witnesses this document and uses it as an input for the attestation to that block.

Reviewer: ~~Individual documents also carry per-document signatures in their `_version` array.~~ This was changed to only sign the blockSignature to save overhead.

> Each indexer signs every block batch it produces. After writing all documents for a block (Block, Transaction, Log, AccessListEntry), the indexer computes a Merkle root over all document CIDs, signs it with its identity key, and writes a `BlockSignature` document. Individual documents also carry per-document signatures in their `_version` array.

Reviewer: ```suggestion
Each indexer signs every block batch it produces. After writing all documents for a block (Block, Transaction, Log, AccessListEntry), the indexer computes a Merkle root over all document CIDs, signs it with its identity key, and writes a `BlockSignature` document. 
```

> Hosts verify incoming `BlockSignature` documents by recomputing the Merkle root and checking the signature against the indexer's known identity. They then create `AttestationRecord` documents that track how many independent indexers produced the same data. The `vote_count` field uses a P-counter CRDT, which merges deterministically across hosts.

Reviewer: ```suggestion
Hosts verify the signature included in the `blockSignature` against the CID for that document. They can then create an `AttestationRecord` (document) for that block to track how many independent indexers produced the same data. The `vote_count` field uses a P-counter CRDT, which tracks the number of indexers which produced that block. 
```

Snapshots bundle multiple blocks into signed files for faster initial sync. The indexer periodically creates `SnapshotSignature` documents whose Merkle root is computed over the per-block `BlockSignature` roots within the range. So you get a two-level Merkle tree: document CIDs roll up into per-block roots, which roll up into per-snapshot roots.

## Registration flows

### Indexer registration

Spans two chains and a relayer:

```mermaid
sequenceDiagram
  participant Op as Indexer operator

>   participant Eth as Ethereum<br/>(Outpost + block extraData)

Reviewer: We removed extra data as `Menna` suggested in this [doc](https://docs.google.com/document/d/1ALIi-FLUv47S4K0YOxZEkp7ZChefSBfIYeH3XmVXVDo/edit?tab=t.0)

This [doc](https://www.notion.so/sourcenetwork/Validator-s-Indexer-Signing-in-Shinzo-3568b3aef35a801a93dbfefd23362175?source=copy_link) outlines the new flow. It is pretty similar.

  Op->>Eth: post EIP-712 attestation<br/>to ChallengeIssuer
  Rel->>Eth: scan extraData,<br/>resolve attestation
  Rel->>Rel: verify EIP-712 sigs
  Rel->>SH: MsgIndexerAssertion
  SH->>SH: record indexer,<br/>derive DID
  SH->>Source: ICA RegisterObject
  Source->>Source: ACP adds DID<br/>to indexer group
```

### Host registration

```mermaid
sequenceDiagram
  participant Op as Operator
  participant SH as ShinzoHub<br/>(Host Registry 0x0211)
  participant Source as SourceHub

  Op->>SH: register(peerKey, nodeKey)
  SH->>SH: verify sigs,<br/>derive DID/PID
  SH->>Source: ICA packet
  Source->>Source: ACP adds DID<br/>to host group
```

### View registration

```mermaid
sequenceDiagram
  participant Dev as Developer
  participant SH as ShinzoHub<br/>(View Registry 0x0210)
  participant Source as SourceHub
  participant Host as Host

  Dev->>SH: viewkit view deploy<br/>(encode VWL, sign EVM tx)
  SH->>SH: decode VWL header,<br/>extract SDL type,<br/>deploy SVS-1 contract
  SH->>Source: ICA RegisterObject
  Source->>Source: ACP registers<br/>view object
  SH-)Host: emit Registered event
  Host->>Host: download bundle,<br/>load WASM lens,<br/>start transforming
```

> ## Transaction flow through a precompile

Reviewer: Is this needed or is it too complex?

Author: Since this is the reference section, I don't think it's too complex. If folks don't wanna know about this, they can just skip over/ignore it. But there absolutely will be nerds out there that want to know what happens via precompile.

The full path of an EVM transaction that hits a registry precompile and triggers a cross-chain ICA call:

1. User sends an EVM tx to a precompile (e.g. `0x0210`).
1. A validator includes the tx in a ShinzoHub block.
1. The EVM calls `precompile.Run()`.
1. The precompile Go code runs five sequential operations:
   1. Decode ABI arguments.
   1. Run business logic (validate, store).
   1. Call the Cosmos keeper (e.g. `RegisterObject`).
   1. The keeper builds an ICA packet and commits it to IBC state.
   1. Emit an EVM log and a Cosmos event.
1. Hermes picks up the ICA packet (async).
1. Hermes submits the packet plus state proof to SourceHub.
1. SourceHub verifies the proof and executes the ACP command.
1. Hermes relays the acknowledgement back to ShinzoHub.

> 

Reviewer: Numbers

Author: https://github.com/shinzonetwork/docs/pull/127/changes#r3194455042
